### PR TITLE
Add more documentation

### PR DESCRIPTION
**Pull request by derekdreery**
I recently submitted a PR to add documentation. This PR adds some more.

Note it won't merge as-is because it relies on PR [#199 in servo/string-cache].

[#199 in servo/string-cache]: https://github.com/servo/string-cache/pull/199

### CODE REVIEW

File: markup5ever/rcdom.rs [hunk 1]
```diff
@@ -11,6 +11,30 @@
...
+//! in HTML would be serialized as
+//!
+//! ```html
+//! <div>text node<span></span></div>
+//! ```
+//!
+//! See the [document object model article on wikipedia][dom wiki] for more information.
+//!
+//! This implementation stores the information associated with each node once, and then hands out
+//! refs to children. The nodes themselves are reference-counted to avoid copy - you can create a
```
Comment (jdm) [inline markup5ever/rcdom.rs]: to avoid copying

File: markup5ever/build.rs [hunk 2]
```diff
@@ -42,26 +42,37 @@ fn main() {
...
 
+    // Create a string cache for namespace urls
     string_cache_codegen::AtomType::new("Namespace", "namespace_url!")
+        .with_macro_doc("Takes a namespace url string and returns its key in a string cache.")
         .atoms(NAMESPACES.iter().map(|&(_prefix, url)| url))
         .write_to(&mut generated)
         .unwrap();
 
-    writeln!(generated, "#[macro_export] macro_rules! ns {{").unwrap();
+    // T
```
[the remark this answers is not in the document]
Comment (derekdreery) [reply markup5ever/build.rs]: oops!

File: markup5ever/build.rs [hunk 3]
```diff
@@ -42,26 +42,37 @@ fn main() {
...
     let local_names = Path::new(&env::var("CARGO_MANIFEST_DIR").unwrap()).join("local_names.txt");
     let mut local_names_atom = string_cache_codegen::AtomType::new("LocalName", "local_name!");
     for line in BufReader::new(File::open(&local_names).unwrap()).lines() {
         let local_name = line.unwrap();
         local_names_atom.atom(&local_name);
         local_names_atom.atom(&local_name.to_ascii_lowercase());
     }
-    local_names_atom.write_to(&mut generated).unwrap();
+    local_names_atom
+        .with_macro_doc("Takes a local name as a string and returns its key in the a string cache.")
```
[the remark this answers is not in the document]
Comment (derekdreery) [reply markup5ever/build.rs]: another oops!

File: markup5ever/interface/mod.rs [hunk 4]
```diff
@@ -119,6 +144,7 @@ impl QualName {
         }
     }
 
+    /// Take a reference as an `ExpandedName`, dropping the unresolved prefix.
```
[the remark this answers is not in the document]
Comment (derekdreery) [reply markup5ever/interface/mod.rs]: I've changed this to "Take a reference of `self`". Do you think this is clear enough, or would you rather something else?

[the remark this answers is not in the document]
Comment (jdm) [reply markup5ever/interface/mod.rs]: I think `Return this qualified name as an `ExpandedName` without the unresolved prefix.` is more clear.

[the remark this answers is not in the document]
Comment (derekdreery) [reply markup5ever/interface/mod.rs]: Ok np. :)

**Comment by derekdreery** [comment]
This is blocked on a new release of string-cache-codegen and dependency bumps.

**Comment by Ygg01** [comment]
@derekdreery Great work, but it seems some upstream changes, kinda broke it.

**Comment by derekdreery** [comment]
Yeah I will update

**Comment by derekdreery** [comment]
I'm trying to tidy up old pull requests. I've rebased off latest master (as of 2018-07-25). I think this is ready for review again.

**Pull request merged**